lib/IRGen/IRGenSIL.cpp

```diff
@@ -4743,26 +4748,66 @@ void IRGenSILFunction::visitDebugValueInst(DebugValueInst *i) {
     SILTy = *MaybeSILTy;
   else
     SILTy = SILVal->getType();
+
   auto RealTy = SILTy.getASTType();
+  if (IsAddrVal && IsInCoro)
+    if (auto *PBI = dyn_cast<ProjectBoxInst>(i->getOperand())) {
+      // Usually debug info only ever describes the *result* of a projectBox
+      // call. To allow the debugger to display a boxed parameter of an async
+      // continuation object, however, the debug info can only describe the box
+      // itself and thus also needs to emit a box type for it so the debugger
+      // knows to call into Remote Mirrors to unbox the value.
+      RealTy = PBI->getOperand()->getType().getASTType();
+      assert(isa<SILBoxType>(RealTy));
+    }
+
+  // Figure out the debug variable type
   if (VarDecl *Decl = i->getDecl()) {
     DbgTy = DebugTypeInfo::getLocalVariable(
         Decl, RealTy, getTypeInfo(SILVal->getType()));
   } else if (!SILTy.hasArchetype() && !VarInfo->Name.empty()) {
-    // Preliminary support for .sil debug information.
+    // Handle the cases that read from a SIL file
     DbgTy = DebugTypeInfo::getFromTypeInfo(RealTy, getTypeInfo(SILTy));
   } else
     return;
 
+  // Since debug_value is expressing indirection explicitly via op_deref,
+  // we're not using either IndirectValue or CoroIndirectValue here.
+  IndirectionKind Indirection = IsInCoro? CoroDirectValue : DirectValue;
+
   // Put the value into a stack slot at -Onone.
   llvm::SmallVector<llvm::Value *, 8> Copy;
-  emitShadowCopyIfNeeded(SILVal, i->getDebugScope(), *VarInfo, IsAnonymous,
-                         Copy);
+  // FIXME: If the SSA value does not have an address type,
+  // shadow copy stack slots will be generated even if the
+  // debug info level is None (i.e. without the `-g` flag).
+  // However, if the SSA value has an address type, its shadow
+  // copy slots will only be generated when debug info level
+  // is greater than None. Not sure if this inconsistency is
+  // intended or not.
+  if (!IsAddrVal)
+    emitShadowCopyIfNeeded(SILVal, i->getDebugScope(), *VarInfo, IsAnonymous,
+                           Copy);
+
   bindArchetypes(DbgTy.getType());
   if (!IGM.DebugInfo)
     return;
 
-  IndirectionKind Indirection =
-    InCoroContext(*CurSILFn, *i) ? CoroDirectValue : DirectValue;
+  // For alloc_stack values, We're currently using llvm.dbg.declare
+  // to model it, so we shouldn't use DW_OP_deref on the associated SSA value.
+  if (isa<AllocStackInst>(SILVal)) {
+    // Drop op_deref
+    // FIXME: Is there a more elegant way to check
+    // if the first element is op_deref and another
+    // utility function to drop it?
+    assert(VarInfo && VarInfo->DIExpr);
+    auto &DIExpr = VarInfo->DIExpr;
+    DIExpr.eraseElement(DIExpr.element_begin());
```

Comment:
just a FYI of how I fixed it: Instead of changing LoadableByAddress Pass, I simply change SILBuiler to teach it not to generate `op_deref` for debug_value if the SILValue is alloc_stack.